src/Features/Core/Portable/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -87,28 +85,74 @@ private string GetRootPathWithGuid_NoLock()
 
         using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
         {
-            _workspace ??= new MetadataAsSourceWorkspace(this, sourceWorkspace.Services.HostServices);
+            if (_workspace is null)
+            {
+                _workspace = new MetadataAsSourceWorkspace(this, sourceWorkspace.Services.HostServices);
+
+                // We're being initialized the first time.  Use this time to clean up any stale metadata-as-source files
+                // from previous VS sessions.
+                CleanupGeneratedFiles(_rootTemporaryPath);
```

Comment:
the logic here inverts, becoming much simpler.  Instead of trying to hook up to shutdown logic to try to clean things at that point, we simply cleanup stale data when we create our first MAS file.